### DISCUSSION TITLE
Is there a way to use Plugins from a private Repository?

### DISCUSSION THREAD

**Discussion by pme123** [com-lihaoyi/mill#2797] [Q&A]
I want to migrate my Ammonite Script to a Mill Plugin (External Module).
The Problem I have is, that it has to be in a private Repo (Not Maven Central).

In Ammonite this is done like this:
```
interp.repositories() ++= Seq(
  MavenRepository
    .of("https://myrepo/releases-maven-local")
    .withCredentials(Credentials.of(sys.env("MVN_REPOSITORY_USERNAME"),
      sys.env("MVN_REPOSITORY_PASSWORD")))
)

@

import $ivy.`valiant:valiant-utils_3:0.4.3 compat`, valiant.utils._

```
Is there a way I can do this with Mill?

I tried also with the Coursier Environment variables. The Repo was listed but the credentials did not to work (401 Unauthorized).
https://get-coursier.io/docs/other-credentials

**Comment by lefou** [comment]
That should be possible. But you need to use a [meta-build](https://mill-build.com/mill/Extending_Mill.html#_the_mill_meta_build), so you can define the repositories in the `mill-build/build.sc`.

Instead of `sys.env`, you need to use [`T.env`](https://mill-build.com/api/latest/mill/api/Ctx.html#env:Map[String,String]) to access the environment variables, otherwise you risk to use stale values.

```scala
// build.sc

import $meta._
```


```scala
// mill-build/build.sc

import mill._
import mill.runner.MillBuildRootModule
import coursier.core.Authentication
import coursier.maven.MavenRepository

object millbuild extends MillBuildRootModule {

  def repositoriesTask: Task[Seq[Repository]] = T.task {
    super.repositoriesTask() ++ 
      MavenRepository(
        "https://myrepo/releases-maven-local",
        Some(Authentication(
          T.env("MVN_REPOSITORY_USERNAME"),
          T.env("MVN_REPOSITORY_PASSWORD")
        ))
      )
  }
}
```

_Edit: Corrected path and pseude code, based on the feedback._

**Comment by pme123** [comment]
Thanks that worked - there where some problems with your snippets though. 
- it is mill-build/build.sc NOT meta-build/build.sc
- The MavenRepository is something like 
```
import coursier.core.Authentication
import coursier.maven.MavenRepository

MavenRepository(
 mavenRepoStr,
 Some(Authentication(user, pwd))
)
```

**Reply by lefou** [reply]
Thanks for the feedback. I just copied your example and edited it a bit, but glad it helped in the end. I guess I had to make it more clear that it's preude-code. ;)

I will update my answer to make it easier for others to reuse it.
